rfcs/2021-10-icons-in-core-DS.md

```diff
@@ -0,0 +1,102 @@
+# RFC: Standardizing Icons in the Core DS
+
+**Date**: October 01, 2021
+**Status**: `Review`
+
+## Problem
+
+The core DS does not have a consistent and easy way for consuming applications to import SVG icons. Additionally, because of the way some icons are used in DS components, it is difficult for consuming applications to change styles for an icon (like color, size, positioning, etc).
+
+### Background
+
+Currently, icons are stored in three main ways:
+
+1. Exported as an individual React component, but not in a centralized location
+2. Stored in the `design-system/images/` directory as .svg files
+3. Included in a component's markup unable to be re-used individually.
+
+The .svg files in the `images/` directory are referenced in the code in two main ways:
+
+1. In CSS as background images
+2. Using the library `inline-react-svg`, .svg files can be imported and used as a component (Note: this does not work with Typescript components)
+
+## Proposal
+
+Icons in the core DS should be moved to a centralized location. Each icon should be exported individually as a React component and updated with a standard size and consistent accessibilty features. This will provide more flexibility to consuming applications by allowing them to import individual SVG icons, adjust the styling on them with more ease, and create icon consistency among products.
+
+This ideal state can be achieved in a few phases:
+
+1. [Create icon components](#phase-1-create-icon-components)
+2. [Clean & improve markup](#phase-2-clean-and-improve-svg-markup)
+3. [Update icon references in components](#phase-3-update-component-icon-references)
+4. [Update icon references in CSS](#phase-4-update-css-icon-references)
+5. [Beyond](#beyond)
+
+### Phase 1: Create Icon Components
+
+In this phase:
+
+- Make `Icons` directory in the design-system component folder
+- Create a component for each icon in the `images` directory
+- Move the `ClearIcon` and `ClearIconAlternate` components into new directory
+- Move inlined SVG markup for `<` & `>` from Pagination to new directory. Consider consolidating to one Icon with variations.
+- Ensure all icons are exported
+- Create base icon component styles to be re-used in each component
+- Update doc site with component details
+
+Exporting each SVG icon individually allows consuming applications to only import the icons they are using.
+
+#### Considerations
+
+There was an idea to create one `Icon` component and pass in a `name` prop that would return the proper svg markup. However, this approach would not work well for child design systems and could inflate bundle sizes.
```

Comment:
* I think the biggest difference in using one component vs may is that is one component, you would be building out a single `Icon` component that basically has a giant conditional statement to figure out which SVG to load. This wouldn't be so bad with our current amount of icons, but if we continue to grow our icon library, it would cause a bunch of SVG markup to be part of the component when a team may only really need one icon and a small amount of markup. Additionally, the conditional to determine the markup is part of the component, so it would be difficult for a child system to use this `Icon` component with a child DS-specific icon.
* In my opinion, I don't think either is "easier" or "harder". Like mentioned before, having only one icon could inflate JS bundle sizes unnecessarily.
* If the design systems were unified, the one icon approach _could_ be extended because it would have easier access to child DS files, and therefore include them as part of the large component conditional. I still don't think that is the way to go because then you run the risk of having mgov specific icons as part of the Icon component that gets pulled into hgov and have a bunch of code that is not relevant to one child system

All this being said... We could do both approaches. I still think that individual icons are a scalable path forward for now, but if people want a unified singular `Icon` component, we could make it and reference the individual icon components as a future enhancement 🤷🏻‍♀️